### PR TITLE
Bump bluepy==2.1.0

### PR DESCRIPTION
**Pull request by tomdele**

### CODE REVIEW

**Review by matz-e — DISMISSED** [review]

**Comment by tristan0x** [comment]
retest this please Jenkins

**Comment by tomdele** [comment]
Has something changed with hdf5 or mpi ? 
This is weird Jenkins was not happy with morphio at first and then with libsonata. 

Could it be the : 
```
depends_on('hdf5~mpi')
depends_on('morphio~mpi', type=('build', 'link'), when='@0.1.0:')
depends_on('morphio~mpi@:2.2.1', type=('build', 'link'), when='@0.1.0')
depends_on('morphio~mpi@2.3.9:', type=('build', 'link'), when='@0.2.0:')
```
from py-libsonata/package.py and morpho-kit ?

Or is it a system problem ?

**Comment by tomdele** [comment]
retest this please Jenkins

**Comment by matz-e** [comment]
I think I see now what's going on.

> Has something changed with hdf5 or mpi ?
> This is weird Jenkins was not happy with morphio at first and then with libsonata.

This is partially due to our setup of passing `~mpi` and `+mpi` through everything, and thus being overly zealous. Also, Spack is not fully backtracing when resolving dependencies (this is apparently a really hard problem).

So, I think here, Spack would see our packages first when concretizing in the *past*, but now sees `py-pytables` (in `brainbuilder`, the spec that fails to concretize) first. That package depends only on `hdf5` (`+mpi` implied per default settings), and thus when it encounters `py-libSONATA` or `py-morphio` that want `~mpi`, it'll blow up.

So we can either dutifully add more `+mpi`/`~mpi` switches, or we can drop them. Now that our cluster seems to handle optional MPI better, @pramodk, we should probably ditch that pass-through.

**Comment by tomdele** [comment]
@matz-e, Thanks for the feedback. Should I take actions ? Adding the +mpi in libsonata or something ?

**Comment by matz-e** [comment]
> @matz-e, Thanks for the feedback. Should I take actions ? Adding the +mpi in libsonata or something ?

Let me see with @pramodk…

**Comment by mgeplf** [comment]
@matz-e / @pramodk  Any luck/decision on fixing this?

**Comment by mgeplf** [comment]
@matz-e / @pramodk Sorry to be a nag, but is there any word on a fix for this? The deployment of this is holding up the release of several components.  Thanks.

**Comment by pramodk** [comment]
Me & Matthias checked few bits and I thought this was concluded. Sorry for missing this.

@tomdele : can we discuss this tomorrow on #spack slack channel of HPC?  we can then sync what are all issues.

**Comment by mgeplf** [comment]
@pramodk > can we discuss this tomorrow on #spack slack channel of HPC? we can then sync what are all issues.

Once the sync happens, can we put the conclusion here (or in the FAQ?) in case its something that impacts others?  Based on the PR, I wouldn't expect any issues to arise, but I understand that version changes have wide-ranging impacts.

**Comment by pramodk** [comment]
Sure! I agree.

**Comment by tomdele** [comment]
> @tomdele : can we discuss this tomorrow on #spack slack channel of HPC? we can then sync what are all issues.

Yes sure !

File: deploy/environments/applications.yaml
Comment (matz-e) [inline deploy/environments/applications.yaml:47]: :scream:

Comment (pramodk) [reply deploy/environments/applications.yaml:47]: 🤷‍♂️

**Comment by pramodk** [comment]
Retest this please

**Comment by pramodk** [comment]
Me / Matthias spent quite some time on this today. Here is brief summary:

Here is how versions and dependency looks like into py-bluepy today:

```python
    version('2.1.0', tag='bluepy-v2.1.0')
    version('2.0.0', tag='bluepy-v2.0.0')
    version('0.16.0', tag='bluepy-v0.16.0')
    version('0.14.15', tag='bluepy-v0.14.15')
    version('0.14.14', tag='bluepy-v0.14.14')
    version('0.14.13', tag='bluepy-v0.14.13')
    version('0.14.12', tag='bluepy-v0.14.12')
    version('0.14.8', tag='bluepy-v0.14.8')
    version('0.14.7', tag='bluepy-v0.14.7')
    version('0.14.6', tag='bluepy-v0.14.6')
    version('0.14.5', tag='bluepy-v0.14.5')
    version('0.14.3', tag='bluepy-v0.14.3')
    version('0.14.1', tag='bluepy-v0.14.1')
    version('0.13.5', tag='bluepy-v0.13.5')
    version('0.12.7', tag='bluepy-v0.12.7')

    depends_on('py-setuptools', type=('build', 'run'))

    # the version of bluepy <2 needed multiple dependencies compatible
    # with h5py < 3.0.0 and bluepy > 2 needs h5py > 3.0.0. Hence the
    # following switches :

    # h5py
    depends_on('py-h5py~mpi@3.0.0:', type='run', when='@2.0.0:')
    depends_on('py-h5py~mpi@2.3:2.99', type='run', when='@:1.0.0')

    # neurom
    depends_on('py-neurom@1.6.0:', type='run', when='@2.0.0:')
    depends_on('py-neurom@1.4.18:1.5.99', type='run', when='@:1.0.0')

    # voxcell
    depends_on('py-voxcell@3.0.0', type='run', when='@2.0.0:')
    depends_on('py-voxcell@2.7.4:2.99', type='run', when='@0.14.16:0.16.0')
    depends_on('py-voxcell@:2.7.3', type='run', when='@:0.14.15')

    # bluepysnap
    depends_on('py-bluepysnap@0.9.0:', type='run', when='@2.0.0:')
    depends_on('py-bluepysnap@0.4.1:0.7.1', type='run', when='@:1.0.0')

    # pandas
    depends_on('py-pandas@1.0.0:', type='run', when='@2.0.0:')
    depends_on('py-pandas@0.17:', type='run', when='@:1.0.0')

    # lazy / cached properties (change of backend for caching)
    depends_on('py-cached-property@1.0:', type='run', when='@2.0.0:')
    depends_on('py-lazy@1.0:', type='run', when='@:1.0.0')

    # common to bluepy <2 and >2 versions
    depends_on('py-libsonata@0.1.6:', type='run', when='^python@3.6:')
    depends_on('py-libsonata@0.1.3:0.1.4', type='run', when='^python@:3.5.99')

    # brion changed package name in 3.3.0 and brion bump in bluepy==2.1.0
    depends_on('brion+python@3.1.0:3.2.0', type='run', when='@:2.0.9')
    depends_on('brion+python@3.3.0:', type='run', when='@2.1.0:')

    # TODO: explicit dependency with hdf5 to avoid limitation of concretizer
    depends_on('hdf5~mpi')

    depends_on('py-bluepy-configfile@0.1.11:', type='run')
    depends_on('py-numpy@1.8:', type='run')
    depends_on('py-pylru@1.2:', type='run')

    # for old versions of bluepy <2.0.0
    depends_on('py-enum34@1.0:', type='run', when='^python@:3.3.99')
    depends_on('py-lxml@3.3:', type='run', when='@:1.0.0')
    depends_on('py-pyyaml@3.10:', type='run', when='@:1.0.0')
    depends_on('py-six@1.0:', type='run', when='@:1.0.0')
    depends_on('py-sqlalchemy@1.0:', type='run', when='@:1.0.0')
    depends_on('py-pyrsistent@:0.17', type='run', when='@:1.0.0')
```

If we look at the versions and dependency requirement based on versions, it's apparent that this is becoming more & more complex in this package.

Spack does concretisation of dependencies based on constraints but the concretizer algorithm is greedy and finalizes details of specs before it is certain that will not lead to constraint errors. The concretiser going to be improved but thats something for future. You can see the list of issues / details https://github.com/spack/spack/wiki/Issue-Disambiguation.

I don't want to go into too much details but what particularly caused the issues in this PR is change in brion dependency from:

```
depends_on('brion+python@3.1.0:', type='run')
```

to 

```
    depends_on('brion+python@3.1.0:3.2.0', type='run', when='@:2.0.9')
    depends_on('brion+python@3.3.0:', type='run', when='@2.1.0:')    
```

I have made some changes / hacks in `applications.yaml` but we would **like to avoid this**. It's quite problematic when we have to install this package on external systems and juggling such constraints is time consuming. And also, py-bluepy is quite fundamental dependency of many other packages.

So my question to py-bluepy maintainers is : can we remove the older versions, simplify the dependencies?

**Comment by tomdele** [comment]
First, thanks a lot for your help. 

I totally agree about the over complex package.py but I am not sure on how I can simplify it. Maybe you can help me on this.

2 months ago we introduced the bluepy v2 change which was quite big and API breaking due to 3 major changes: 
- simplification of the API --> removal of a lot of historical dependencies
- python2 removal --> change in the lazy caching lib, removal of `six`, ...
- h5py major version bump from 2 to 3. H5py is used in many tools and this change was breaking for a lot of our bbp tools mainly: neurom, bluepysnap, voxcell. And they are all dependencies of bluepy.

Later on, brion changed name for his python library from `brain` to `brion` (hence this PR) which is also breaking for me in bluepy.

Moreover, bluepy is used in multiple bbp packages and I am not sure they made the bluepy v2 transition yet. Just as an example:
- circuit-analysis, psp-validation are using bluepy<2.0.0
- bglibpy, bbp-analysis, simwriter don't ping any version of bluepy so I guess they are working fine (no idea)
- plus all the tools using these libraries as deps. 
 
So basically, I **think** I still need to have at least 2 versions of bluepy: 
- one with `h5py > 3` (`bluepy>=2.0.0`) 
- one with `h5py<3` (`bluepy==0.16.0`)

Which means I need the switches for the full suite of dependencies (neurom, bluepysnap, voxcell) plus the extra libraries for the version 0.16 I removed or changed in the v2.

So basically, I am not sure I can change anything to the package except removing these lines in the package for the moment: 
```    
depends_on('py-libsonata@0.1.3:0.1.4', type='run', when='^python@:3.5.99')
depends_on('py-voxcell@:2.7.3', type='run', when='@:0.14.15')
```
and all these versions :
```
version('0.14.15', tag='bluepy-v0.14.15')
    version('0.14.14', tag='bluepy-v0.14.14')
    version('0.14.13', tag='bluepy-v0.14.13')
    version('0.14.12', tag='bluepy-v0.14.12')
    version('0.14.8', tag='bluepy-v0.14.8')
    version('0.14.7', tag='bluepy-v0.14.7')
    version('0.14.6', tag='bluepy-v0.14.6')
    version('0.14.5', tag='bluepy-v0.14.5')
    version('0.14.3', tag='bluepy-v0.14.3')
    version('0.14.1', tag='bluepy-v0.14.1')
    version('0.13.5', tag='bluepy-v0.13.5')
    version('0.12.7', tag='bluepy-v0.12.7')
```

Do you have any idea on how to improve the situation ?

**Comment by tomdele** [comment]
I think most of the libs are done with bluepy < 2 and the spack PR are ready or in progress. 

@pramodk / @matz-e Can we remove the bluepy version < 2 from the package.py if some package are still using the dependency explicitly ?

**Comment by tomdele** [comment]
We discussed a bit with @GianlucaFicarelli and we think the safer is:
- merging this PR if possible
- merging https://github.com/BlueBrain/spack/pull/1054 (check if we can remove the bluepy switch in this PR which will remove the brion switch in bluepy if we keep only the version 2.1.0)
- @GianlucaFicarelli has a PR for psp-validation 
- do the circuit analysis spack release
- clean the bluepy package and remove all the versions < 2.1

Does this seems reasonable ?

**Comment by matz-e** [comment]
Seems good from my side.

**Comment by tomdele** [comment]
so no objections if we put this in ready to review and merge ?

**Review by matz-e — APPROVED** [review]
Go ahead!

**Comment by tomdele** [comment]
Thank you !

**Comment by GianlucaFicarelli** [comment]
Unfortunately `bluepy` 2.1 is not available in unstable today. It seems that the jenkins build failed [here](https://bbpcode.epfl.ch/ci/view/Lastest%20Builds/job/hpc.spack-deployment/3192/consoleFull).
The build is marked as `Unstable` and in the logs there are some errors, including
`04:57:02  /usr/bin/tar: Error is not recoverable: exiting now`.
Could it be a temporary error, or need more investigation?
cc @tomdele @matz-e

**Comment by tomdele** [comment]
I had the same error before : 
https://github.com/BlueBrain/spack/pull/1033#issuecomment-772430826

And I think it was due to the system. I am not sure if we can just rerun the deployment part from jenkins...

**Comment by matz-e** [comment]
> It seems that the jenkins build failed here: https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.spack-deployment/detail/hpc.spack-deployment/3192/tests

FTFY. Embrace Blue Ocean :)

I'm going to retrigger it.

**Comment by GianlucaFicarelli** [comment]
`py-blueby` 2.1.0 is still not available today because the last build failed again :( but with a different error, something related to gcc/intel compilers.
I suppose we need to add another `^freetype%gcc` for `psp-validation`
```
04:40:20  ==> Error: Conflicts in concretized spec "psp-validation@0.3.3%gcc@9.3.0 arch=linux-rhel7-x86_64/2lh5dqa"
04:40:20  
04:40:20  List of matching conflicts for spec:
04:40:20  
04:40:20      freetype@2.10.1%intel@19.1.2.254 arch=linux-rhel7-x86_64
04:40:20          ^bzip2@1.0.8%intel@19.1.2.254+shared arch=linux-rhel7-x86_64
04:40:20          ^libpng@1.6.37%gcc@9.3.0 arch=linux-rhel7-x86_64
04:40:20              ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-rhel7-x86_64
04:40:20          ^pkgconf@1.7.3%intel@19.1.2.254 arch=linux-rhel7-x86_64
04:40:20  
04:40:20  1. "%intel" conflicts with "freetype@2.8:" [freetype-2.8 and above cannot be built with icc (does not support __builtin_shuffle)]
```

EDIT: opened https://github.com/BlueBrain/spack/pull/1080

**Pull request merged**
